lib/pbench/server/database/database.py

```diff
@@ -0,0 +1,44 @@
+from sqlalchemy import create_engine
+from sqlalchemy.orm import scoped_session, sessionmaker
+from sqlalchemy.ext.declarative import declarative_base
+from pbench.server import NoOptionError, NoSectionError
+
+
+class Database:
+    # Create declarative base model that our model can inherit from
+    Base = declarative_base()
+    # Initialize the db scoped session
+    db_session = scoped_session(sessionmaker(autocommit=False, autoflush=False))
+
+    @staticmethod
+    def get_engine_uri(config, logger):
+        try:
+            psql = config.get("Postgres", "db_uri")
+            return psql
+        except (NoSectionError, NoOptionError):
+            logger.error(
+                "Failed to find an [Postgres] section" " in configuration file."
+            )
+            return None
+
+        # return f"postgresql+{psql_driver}://{psql_username}:{psql_password}@{psql_host}:{psql_port}/{psql_db}"
+
+    @staticmethod
+    def init_engine(server_config, logger):
+        try:
+            return create_engine(Database.get_engine_uri(server_config, logger))
+        except Exception:
+            logger.exception("Exception while creating a sqlalchemy engine")
+            return None
+
+    @staticmethod
+    def init_db(server_config, logger):
+        # Make sure all the models are imported before this function gets called
+        # so that they will be registered properly on the metadata. Otherwise
+        # metadata will not have any tables and create_all functionality will do nothing
+
+        Database.Base.query = Database.db_session.query_property()
+
+        engine = create_engine(Database.get_engine_uri(server_config, logger))
+        Database.Base.metadata.create_all(bind=engine)
+        Database.db_session.configure(bind=engine)
```

Comment:
Some items which came up in #2101 which Dave indicated had originated here; these don't need to hold up the merge, but it would be good if they were addressed in a subsequent issue/PR.

First, is it possible for the code to determine whether all the models have been imported when we invoke this function?  Or, failing that, can the code determine that there are no tables or that the `create_all()` function accomplished nothing?  It would be much better to produce a run-time failure with a clear diagnostic than it would be to run a little further and then die cryptically with a database error about referencing a non-existent table.

Second, the assignments here appear to be addressing class members, when there is also class initialization being done at lines 9-11.  It would be better if class initialization were all done in one place, unless there are compelling reasons to split it.  (I understand that we don't have a logger at line 9, and that limits how/when/where the engine can be created and everything which depends on it, so, unless there's a compelling need to have class initialization done at line 9, perhaps those assignments should be moved to here.)  (Also, should this be a `@classmethod` instead of a `@staticmethod`, since it seeks to modify the class members?)

Reply:
About your first concern, all the models should get imported before we call the database `init` function inside `create_app` on line 120 in the `lib/pbench/server/api/__init__.py` file so we don't need to explicitly import all the models again. 

About the second point, the variables on line 9 and 11 are defined as a class member so they can be easily imported in our model definition without creating a separate object. The reason I created them as a class variable instead of object variable is because they need to be same and consistent across all the models we import them in and to make sure they are using the same static database session. Creating an object variable will require us to keep passing the same object instance everywhere.

Reply:
My first concern was principally keying on the code comment.  When a comment says "Make sure that...", it suggests that there is an opportunity for a silent failure which will create a pitfall for a future developer.  It would be much better if some piece of code (either the server code itself, a build procedure, or a unit test) could do the check instead of relying on the developer to see, understand, and act on the code comment.  If as you say, the models _are_ already being imported, then maybe the code comment should be removed or modified to reflect what you said in your comment here.

As for my second point, that followed from a misapprehension of how the database models work.  I think I'm mostly straightened out on that (although, I'm not really excited by the new-found knowledge...), so I think we can ignore that part of my comment, now.

Reply:
Another area that could use some cleanup in a follow-on PR. (I assume that you'll move on to finish your metadata PR #2049, so that's a possibility if not handled separately.)

Operationally; if we can do anything here to ensure/detect that everything's good, that would be excellent. If not, let's at least clean up the comment to describe the assumptions and current implementation, providing a simple and actionable warning about the impact of changes on those assumptions ... rather than a vague and disturbing warning with no actionable mitigation.